Sources/Swiftly/Use.swift

```diff
@@ -46,43 +53,187 @@ internal struct Use: SwiftlyCommand {
+        if let selectedVersion = selectedVersion {
+            message += " (was \(selectedVersion.name))"
         }
 
         SwiftlyCore.print(message)
     }
+
+    internal static func findNewVersionFile() -> URL? {
+        var cwd = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
+
+        while cwd.path != "" && cwd.path != "/" {
+            guard FileManager.default.fileExists(atPath: cwd.path) else {
+                break
+            }
+
+            let gitDir = cwd.appendingPathComponent(".git")
+
+            if FileManager.default.fileExists(atPath: gitDir.path) {
+                return cwd.appendingPathComponent(".swift-version")
+            }
+
+            cwd = cwd.deletingLastPathComponent()
+        }
+
+        return nil
+    }
+}
+
+public enum ToolchainSelectionResult {
+    case globalDefault
+    case swiftVersionFile(URL, ToolchainSelector?, Error?)
+}
+
+/// Returns the currently selected swift toolchain, if any, with details of the selection.
+///
+/// The first portion of the returned tuple is the version that was selected, which
+/// can be nil if none can be selected.
+///
+/// Selection of a toolchain can be accomplished in a number of ways. There is the
+/// the configuration's global default 'inUse' setting. This is the fallback selector
+/// if there are no other selections. The returned tuple will contain the default toolchain
+/// version and the result will be .default.
+///
+/// A toolchain can also be selected from a `.swift-version` file in the current
+/// working directory, or an ancestor directory. If it successfully selects a toolchain
+/// then the result will be .swiftVersionFile with the URL of the file that made the
+/// selection and the first item of the tuple is the selected toolchain version.
+///
+/// However, there are cases where the swift version file fails to select a toolchain.
+/// If such a case happens then the toolchain version in the tuple will be nil, but the
+/// result will be .swiftVersionFile and a detailed error about the problem. This error
+/// can be thrown by the client, or ignored.
+public func selectToolchain(config: inout Config, globalDefault: Bool = false, install: Bool = false) async throws -> (ToolchainVersion?, ToolchainSelectionResult) {
+    if !globalDefault {
+        var cwd = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
+
+        while cwd.path != "" && cwd.path != "/" {
+            guard FileManager.default.fileExists(atPath: cwd.path) else {
+                break
+            }
+
+            let svFile = cwd.appendingPathComponent(".swift-version")
+
+            if FileManager.default.fileExists(atPath: svFile.path) {
+                let contents = try? String(contentsOf: svFile, encoding: .utf8)
+
+                guard let contents = contents else {
+                    return (nil, .swiftVersionFile(svFile, nil, Error(message: "The swift version file could not be read: \(svFile)")))
```

Comment:
The error is part of the tuple here because in some contexts the error is ignored, while others it is thrown.

For example, when `swiftly use` is displaying the currently selected toolchain, or it is trying to find the swift version file to update with the used version it doesn't care if the file is well-formed, or not matching a toolchain that is currently installed. Whereas in other cases like the proxies or the `swiftly run` that will interfere with the ability to run the command against a toolchain.